lib/src/base_validator.dart

```diff
@@ -0,0 +1,29 @@
+/// Base class for all validators.
+abstract class BaseValidator<T> {
+  /// {@template base_validator_error_text}
+  /// The error message returned if the value is invalid.
+  /// {@endtemplate}
+  final String errorText;
+
+  const BaseValidator({required this.errorText});
+
+  /// Validates the given value.
+  String? validate(T? valueCandidate);
+
+  /// Checks if the given value is valid on basics:
+  ///
+  /// - If the value is `null`.
+  /// - If the value is a `String`, `Iterable`, or `Map` and is empty.
+  /// - If the value passes the custom validation made by `validate`.
+  ///
+  /// Returns the error message if the value is invalid.
+  String? call(T? valueCandidate) {
```

Comment:
`call` maybe doesn't describe it enough. Maybe we can call it: `validateValue` or `validateNullOrEmpty`?